rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -179,4 +186,170 @@ impl MultisigCheckpointSyncer {
+        for validator in validators.iter() {
+            let address = H160::from(*validator);
+            if let Some(checkpoint_syncer) = self.checkpoint_syncers.get(&address) {
+                // Gracefully handle errors getting the latest_index
+                match checkpoint_syncer.latest_index().await {
+                    Ok(Some(index)) => {
+                        trace!(?address, ?index, "Validator returned latest index");
+                        latest_indices.push(index);
+                    }
+                    err => {
+                        debug!(?address, ?err, "Failed to get latest index from validator");
+                    }
+                }
+            }
+        }
+        debug!(latest_indices=?latest_indices, "Fetched latest indices from checkpoint syncers");
+
+        if latest_indices.is_empty() {
+            debug!("No validators returned a latest index");
+            return Ok(None);
+        }
+
+        // Sort in descending order. The n'th index will represent
+        // the highest index for which we (supposedly) have (n+1) signed checkpoints
+        latest_indices.sort_by(|a, b| b.cmp(a));
+        if let Some(&highest_quorum_index) = latest_indices.get(threshold - 1) {
+            // The highest viable checkpoint index is the minimum of the highest index
+            // we (supposedly) have a quorum for, and the maximum index for which we can
+            // generate a proof.
+            let start_index = highest_quorum_index.min(maximum_index);
+            if minimum_index > start_index {
+                debug!(%start_index, %highest_quorum_index, "Highest quorum index is below the minimum index");
+                return Ok(None);
+            }
+            for index in (minimum_index..=start_index).rev() {
+                if let Ok(Some(checkpoint)) =
+                    self.fetch_checkpoint(validators, threshold, index).await
+                {
+                    return Ok(Some(checkpoint));
+                }
+            }
+        }
+        debug!("No checkpoint found in range");
+        Ok(None)
+    }
+
+    /// Fetches a MultisigSignedCheckpointWithMessageId if there is a quorum.
+    /// Returns Ok(None) if there is no quorum.
+    #[instrument(err, skip(self))]
+    pub async fn fetch_checkpoint(
+        &self,
+        validators: &[H256],
+        threshold: usize,
+        index: u32,
+    ) -> Result<Option<MultisigSignedCheckpoint<CheckpointWithMessageId>>> {
+        // Keeps track of signed validator checkpoints for a particular root.
+        // In practice, it's likely that validators will all sign the same root for a
+        // particular index, but we'd like to be robust to this not being the case
+        let mut signed_checkpoints_per_root: HashMap<
+            H256,
+            Vec<SignedCheckpointWithSigner<CheckpointWithMessageId>>,
+        > = HashMap::new();
+
+        for validator in validators.iter() {
```

Comment:
I think this is valid
```suggestion
        for validator in validators {
```